tka/sig.go

```diff
@@ -42,17 +48,47 @@ func (s SigKind) String() string {
 type NodeKeySignature struct {
 	// SigKind identifies the variety of signature.
 	SigKind SigKind `cbor:"1,keyasint"`
-	// Pubkey identifies the public key which is being certified.
+	// Pubkey identifies the public key which is being authorized.
 	Pubkey []byte `cbor:"2,keyasint"`
 
 	// KeyID identifies which key in the tailnet key authority should
 	// be used to verify this signature. Only set for SigDirect and
 	// SigCredential signature kinds.
 	KeyID []byte `cbor:"3,keyasint,omitempty"`
 
-	// Signature is the packed (R, S) ed25519 signature over the rest
-	// of the structure.
+	// Signature is the packed (R, S) ed25519 signature over all other
+	// fields of the structure.
 	Signature []byte `cbor:"4,keyasint,omitempty"`
+
+	// Nested describes a NodeKeySignature which authorizes the node-key
+	// used as Pubkey. Only used for SigRotation signatures.
+	Nested *NodeKeySignature `cbor:"5,keyasint,omitempty"`
+
+	// RotationPubkey specifies the ed25519 public key which may sign a
+	// SigRotation signature, which embeds this one.
+	//
+	// Intermediate SigRotation signatures may omit this value to use the
+	// parent one.
+	RotationPubkey []byte `cbor:"6,keyasint,omitempty"`
+}
+
+// rotationPublic returns the public key which must sign a SigRotation
+// signature that embeds this signature, if any.
+func (s NodeKeySignature) rotationPublic() (pub ed25519.PublicKey, ok bool) {
+	if len(s.RotationPubkey) > 0 {
+		return ed25519.PublicKey(s.RotationPubkey), true
+	}
+
+	switch s.SigKind {
+	case SigRotation:
+		if s.Nested == nil {
+			return nil, false
+		}
+		return s.Nested.rotationPublic()
```

Comment:
will this be optimized into tail recursion? Otherwise it might be better to make this a loop.

Reply:
I don't think so, but I'm also hoping to keep nesting a few levels deep so hopeful optimizations like an array or a loop wont be necessary.